cudax/include/cuda/experimental/__device/all_devices.cuh

```diff
@@ -0,0 +1,244 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of CUDA Experimental in CUDA C++ Core Libraries,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDAX__DEVICE_ALL_DEVICES
+#define _CUDAX__DEVICE_ALL_DEVICES
+
+#include <cuda/__cccl_config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <cuda/std/__cccl/attributes.h>
+#include <cuda/std/__cuda/api_wrapper.h>
+#include <cuda/std/atomic>
+#include <cuda/std/cassert> // IWYU pragma: keep (for assert)
+#include <cuda/std/iterator> // IWYU pragma: keep (for random_access_iterator_tag)
+
+#include <cuda/experimental/__device/device.cuh>
+
+namespace cuda::experimental
+{
+namespace __detail
+{
+//! @brief A random-access range of all available CUDA devices
+class all_devices
+{
+public:
+  using size_type = _CUDA_VSTD::size_t;
+  struct iterator;
+  using const_iterator = iterator;
+
+  all_devices() = default;
```

Comment:
I am a bit torn here, if I am reading this correctly we have a mutable atomic member just to ensure that the struct is constexpr?

Otherwise we could just move the query into the constructor and be safe?

Reply:
can i be making calls to the cuda runtime before `main`? i thought that was a no-no.

Reply:
urgh